### ISSUE TITLE
Invalid root= line in OEM image

### ISSUE THREAD

**Issue by Vogtinator** [OSInside/kiwi#1287]
As observed for Virtualization:Appliances:Images:openSUSE-Tumbleweed/kiwi-templates-JeOS:RaspberryPi, the generated grub configuration is invalid, it passes `root=/dev/mapper/loop0p3` to the kernel.

This probably happened because the `/dev/block/by-uuid/...` symlink was not available at the time `grub2-mkconfig` was called.

**Comment by schaefi** [comment]
This also happens for other builds and seems to be a problem if the build happens in obs.
For example see: Virtualization:Appliances:Images:Testing_x86:suse/test-image-vmx

If you go and build it locally the issue went away.

I think there is little we can do as it seems to be an environment issue. I'm a bit concerned because this behavior looks new to me, we didn't saw it before. It's also very generic and happens everywhere, also for other distros: e.g Virtualization:Appliances:Images:Testing_x86:fedora/test-image-iso-oem-vmx

So do you have any idea how we can influence this on the obs worker level ?

**Comment by Vogtinator** [comment]
The `/dev/block/by-uuid/...` links are created by udev, which isn't running in OBS workers. Just creating the links manually should be enough.

**Comment by schaefi** [comment]
no udev in the worker can cause all kinds of other problems, e.g for building lvm images, or crypto or any other type of image that uses an on top system that creates udev rule based nodes. Why not simply run udev in the workers ?

**Comment by schaefi** [comment]
at least for building images, or give us an opportunity to configure it through prjconf or similar. I don't think code in kiwi that manually creates links makes sense

**Comment by schaefi** [comment]
that would just be yet another source for errors as it should be managed by udev

**Comment by Vogtinator** [comment]
There would have to be some code in kiwi that ensures that udev rules ran after modifying a loop device anyway, otherwise it'd be full of race conditions

**Comment by Vogtinator** [comment]
About running udev in obs workers, that's something to discuss with the OBS team.

**Comment by schaefi** [comment]
>There would have to be some code in kiwi that ensures that udev rules ran after modifying a loop >device anyway, otherwise it'd be full of race conditions

no why ?, all loop tools are called either in sync mode returning the device or kiwi actively waits for the udev queue to be done. If it would be full of race conditions we would see them every day.

> About running udev in obs workers, that's something to discuss with the OBS team.

yes please, whom should I contact about it ?

It seems to be a pretty severe problem. We changed everything to use grub2-mkconfig on your request and you had good arguments to do it, but if obs workers comes with an environment that does not allow its execution we are in big trouble

@davidcassany please hold back any submission to SLE15 as I will hold back mine to SLE12. I will also check if this problem also exists when building for SLE.

Thanks

**Comment by schaefi** [comment]
Also happens on SLE: e.g QA:Maintenance:Images:12-SP5:kiwi-ng-testing/test-image-vmx

**Comment by schaefi** [comment]
ok so this one is really really bad.

**Comment by schaefi** [comment]
I revert the kiwi in QA:Maintenance. Thanks Fabian for reporting this prior any SLE submission. However Factory image building in obs is completely broken now :(

**Comment by Vogtinator** [comment]
> > There would have to be some code in kiwi that ensures that udev rules ran after modifying a loop >device anyway, otherwise it'd be full of race conditions
> 
> no why ?, all loop tools are called either in sync mode returning the device

That means the kernel created the device, not that udev made the links accessible.
The more critical part is the `by-uuid` link, which can only be done after the FS creation.

> or kiwi actively waits for the udev queue to be done. If it would be full of race conditions we would see them every day.

I didn't see any call to `udevadm settle` during build - otherwise it would just fail in OBS already as it's not installed.

> > About running udev in obs workers, that's something to discuss with the OBS team.
> 
> yes please, whom should I contact about it ?

@adrianschroeter, @mlschroe, at openSUSE/obs-build.

> It seems to be a pretty severe problem. We changed everything to use grub2-mkconfig on your request and you had good arguments to do it, but if obs workers comes with an environment that does not allow its execution we are in big trouble

It shouldn't be too hard to fix that environment (and add some sanity checks). So far, the grub2-mkconfig support worked out well. I wonder how/why it broke now... Maybe it's a totally different issue? Unfortunately I don't know how to debug this easily.

> @davidcassany please hold back any submission to SLE15 as I will hold back mine to SLE12. I will also check if this problem also exists when building for SLE.
> 
> Thanks

**Comment by adrianschroeter** [comment]
On Mittwoch, 4. Dezember 2019, 10:12:40 CET Fabian Vogt wrote:
> > > There would have to be some code in kiwi that ensures that udev rules ran after modifying a loop >device anyway, otherwise it'd be full of race conditions
> > 
> > no why ?, all loop tools are called either in sync mode returning the device
> 
> That means the kernel created the device, not that udev made the links accessible.
> The more critical part is the `by-uuid` link, which can only be done after the FS creation.
> 
> > or kiwi actively waits for the udev queue to be done. If it would be full of race conditions we would see them every day.
> 
> I didn't see any call to `udevadm settle` during build - otherwise it would just fail in OBS already as it's not installed.
> 
> > > About running udev in obs workers, that's something to discuss with the OBS team.

this would be distribution specific. So it must be done via packages.

IIRC our udev package used to have a %post script executing udevd ....




-- 

Adrian Schroeter <adrian@suse.de>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

**Comment by schaefi** [comment]
>I didn't see any call to `udevadm settle` during build - otherwise it would just fail in OBS already as >it's not installed.

as I said we call tools in sync mode (-s). per boot code see: modules.d/99kiwi-lib/kiwi-lib.sh

**Comment by Vogtinator** [comment]
> as I said we call tools in sync mode (-s)

This would be mostly relevant for `mkfs.btrfs` and such - those don't have such an option and I also don't know how they could wait for udev

**Comment by schaefi** [comment]
>this would be distribution specific. So it must be done via packages.
>IIRC our udev package used to have a %post script executing udevd ....

what does that mean ? is there a way for obs workers to run udevd or not ?

**Comment by adrianschroeter** [comment]
On Mittwoch, 4. Dezember 2019, 10:22:02 CET Marcus Schäfer wrote:
> >this would be distribution specific. So it must be done via packages.
> >IIRC our udev package used to have a %post script executing udevd ....
> 
> what does that mean ? is there a way for obs workers to run udevd or not ?

sure, the question is just who should start it.

The build script is not the best place for this, since the way a daemon starts
may change. So it should be done by the distribution itself, means by some rpm.

-- 

Adrian Schroeter <adrian@suse.de>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

**Comment by schaefi** [comment]
> This would be mostly relevant for mkfs.btrfs

never had any issue to mount and use a filesystem after the mkfs.XX command returned. Anyway if you think there is a race condition let's discuss this in another thread. I'm mostly concerned about grub2-mkconfig not working in a obs worker and would like to find a solution

**Comment by schaefi** [comment]
> means by some rpm

an rpm has to start udev. So how do you handle other service types, e.g systemd ?

**Comment by adrianschroeter** [comment]
On Mittwoch, 4. Dezember 2019, 10:25:11 CET Marcus Schäfer wrote:
> > means by some rpm
> 
> an rpm has to start udev. So how do you handle other service types, e.g systemd ?

it is not running. Why would a build process need systemd?

-- 

Adrian Schroeter <adrian@suse.de>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

**Comment by Vogtinator** [comment]
> > This would be mostly relevant for mkfs.btrfs
>
> never had any issue to mount and use a filesystem after the mkfs.XX command returned.

That's not the issue - the `/dev/block/by-uuid/...` links based on the FS UUID are created by udev async. So only after `mkfs` + `udevadm settle` that link, which `grub2-mkconfig` relies on, is guaranteed to exist.

I'm currently trying to find out which change caused the regression. It can't be a recent one, otherwise we'd have noticed it earlier...

**Comment by adrianschroeter** [comment]
I wonder why the build environement esp. the available storage devices should influence the build result actually. The build environment would not exist when the image gets deployed, so it shouldn't matter concept wise, no?

**Comment by schaefi** [comment]
> it is not running. Why would a build process need systemd?

just thought a worker is a plain service VM, running an OS with all default services to be near to a local build environment. The image build process doesn't need systemd, I just though obs might have some api to run services

**Comment by Vogtinator** [comment]
kiwi calls `grub2-mkconfig` chrooted during build, which runs this snippet in `/etc/grub.d/10_linux`:

```
# btrfs may reside on multiple devices. We cannot pass them as value of root= parameter
# and mounting btrfs requires user space scanning, so force UUID in this case.
if [ "x${GRUB_DEVICE_UUID}" = "x" ] || [ "x${GRUB_DISABLE_LINUX_UUID}" = "xtrue" ] \
    || ! test -e "/dev/disk/by-uuid/${GRUB_DEVICE_UUID}" \
    || ( test -e "${GRUB_DEVICE}" && uses_abstraction "${GRUB_DEVICE}" lvm ); then
  LINUX_ROOT_DEVICE=${GRUB_DEVICE}
else
  LINUX_ROOT_DEVICE=UUID=${GRUB_DEVICE_UUID}
fi
```

`GRUB_DEVICE_UUID` is read from `blkid`, but the `/dev/disk/by-uuid/${GRUB_DEVICE_UUID}` is created by udev. If it does not exist, it silently falls back to using the currently active device of the `/` mountpoint.

**Comment by schaefi** [comment]
> I'm currently trying to find out which change caused the regression.

This was because of the fix in #1284. Before that we had two root= lines, one always correct one by kiwi and the other created by grub2-mkconfig. Since the fix prevented the first to be written the issue came up. I think it was always there we just didn't recognized it because the images always booted

**Comment by schaefi** [comment]
I have an idea for a solution... give me some minutes
